### PR TITLE
Refactor reporting to use primitives (classes)

### PR DESCRIPTION
**Pull request by Rezney**
Also:

- add tags
- remove renderers
- extend remediations for Insights use-case

kudos to @shaded-enmity

### CODE REVIEW

File: repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py [hunk 1]
```diff
@@ -22,14 +22,21 @@ def process(self):
...
+                    create_report(
+                        reporting.Title('Btrfs has been removed from RHEL8'),
+                        reporting.Summary('The Btrfs file system was introduced as Technology Preview with the initial
+                            'release of Red Hat Enterprise Linux 6 and Red Hat Enterprise Linux 7. As of versions 6.6 '
+                            'and 7.4 this technology has been deprecated and removed in RHEL8.'),
+                        reporting.ExternalLink(
+                            title='Considerations in adopting RHEL 8 - btrfs has been removed.',
+                            url='https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/considerations_in_adopting_rhel_8/file-systems-and-storage_considerations-in-adopting-rhel-8#btrfs-has-been-removed_file-systems-and-storage'
+                        )
+                        reporting.Severity('high'),
```
Comment (asilveir) [inline repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py]:
```suggestion
                        reporting.Severity(reporting.Severities.HIGH),
```
Any chance to add this?

Comment (asilveir) [reply repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py]: BTW, very nice work congrats.

Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py]: Original idea is @shaded-enmity 's :)

Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py]: was thinking about the "severity stuff" too...

Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py]: severities implemented. Thanks for the suggestions.

File: repos/system_upgrade/el7toel8/actors/authselectcheck/actor.py [hunk 2]
```diff
@@ -83,14 +83,24 @@ def produce_authconfig_configuration(self, model, command):
...
 
-        report_generic(
-            title='Authselect will be used to configure PAM and nsswitch.conf.',
-            summary='There is a new tool called authselect in RHEL8 that '
-                    'replaced authconfig. The upgrade process detected '
-                    'that authconfig was used to generate current '
-                    'configuration and it will automatically convert it '
-                    'to authselect. Authselect call is: {}'.format(command)
-        )
+        create_report([
```
Comment (jmikovic) [inline repos/system_upgrade/el7toel8/actors/authselectcheck/actor.py]: as discussed offline, synchronize using either list or objects as argument for create_report

File: repos/system_upgrade/el7toel8/actors/authselectcheck/actor.py [hunk 3]
```diff
@@ -99,39 +109,68 @@ def produce_current_configuration(self, model):
...
+                'to find an authselect profile that would be equivalent '
+                'to your current configuration. Therefore your '
+                'configuration will be left intact.'
+            ),
+            reporting.Tags([
+                reporting.Tags.AUTHENTICATION,
+                reporting.Tags.SECURITY,
+                reporting.Tags.TOOLS
+            ]),
+            reporting.Severity(reporting.Severity.LOW)
```
Comment (jmikovic) [inline repos/system_upgrade/el7toel8/actors/authselectcheck/actor.py]:
```suggestion
            reporting.Severity(reporting.Severity.INFO)
```

File: repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py [hunk 4]
```diff
@@ -22,14 +22,21 @@ def process(self):
...
-                        title='Btrfs removed in the next major version',
-                        summary='The Btrfs file system was introduced as Technology Preview with the initial release '
-                                'of Red Hat Enterprise Linux 6 and Red Hat Enterprise Linux 7. As of versions 6.6 and '
-                                '7.4 this technology has been deprecated and will be removed in next major version.',
-                        remediation='Please consider migrating your Btrfs mount point(s) to a different filesystem '
-                                    'before next upgrade attempt. If no Btrfs filesystem is in use, please unload '
-                                    'btrfs kernel module running "# rmmod btrfs".',
-                        severity='high',
-                        flags=['inhibitor'])
+                    create_report(
```
Comment (jmikovic) [inline repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py]: missing remediation

File: repos/system_upgrade/el7toel8/actors/checksystemarch/actor.py [hunk 5]
```diff
@@ -21,8 +21,10 @@ class CheckSystemArch(Actor):
 
     def process(self):
         if platform.machine() != 'x86_64':
-            report_generic(
-                title='Unsupported arch',
-                summary='Upgrade process is only supported on x86_64 systems.',
-                severity='high',
-                flags=['inhibitor'])
+            create_report([
+                reporting.Title('Unsupported arch'),
```
Comment (shaded-enmity) [inline repos/system_upgrade/el7toel8/actors/checksystemarch/actor.py]:
```suggestion
                reporting.Title('Unsupported architecture'),
```

File: repos/system_upgrade/el7toel8/actors/createresumeservice/actor.py [hunk 6]
```diff
@@ -35,8 +35,12 @@ def process(self):
...
-            summary='{} enabled as oneshot systemd service to resume Leapp execution '
-                    'after reboot.'.format(service_name),
-            severity='low')
+        create_report([
+            reporting.Title('Leapp resume systemd service enabled'),
+            reporting.Summary(
+                '{} enabled as oneshot systemd service to resume Leapp execution '
+                'after reboot.'.format(service_name)
+            ),
+            reporting.Severity(reporting.Severity.LOW),
```
Comment (shaded-enmity) [inline repos/system_upgrade/el7toel8/actors/createresumeservice/actor.py]:
```suggestion
            reporting.Severity(reporting.Severity.INFO),
```

File: repos/system_upgrade/el7toel8/actors/inhibitwhenluks/actor.py [hunk 7]
```diff
@@ -21,12 +21,16 @@ def process(self):
...
-                                    'the system',
-                        severity='high',
-                        flags=['inhibitor'])
+                    create_report([
+                        reporting.Title('LUKS encrypted partition detected'),
+                        reporting.Summary('Upgrading system with encrypted partitions is not supported'),
+                        reporting.Severity(reporting.Severity.HIGH),
+                        reporting.Tags([reporting.Tags.BOOT, reporting.Tags.ENCRYPTION]),
+                        reporting.Flags([reporting.Flags.INHIBITOR]),
+                        reporting.Remediation(
```
Comment (shaded-enmity) [inline repos/system_upgrade/el7toel8/actors/inhibitwhenluks/actor.py]: I'd get rid of this hint, it doesn't seem to provide any info at all.

File: repos/system_upgrade/el7toel8/actors/ipuworkflowconfig/libraries/library.py [hunk 8]
```diff
@@ -31,9 +32,11 @@ def get_os_release(path):
...
-        reporting.report_generic(
-            title='Error while collecting system OS facts',
-            summary=str(e),
-            severity='high',
-            flags=['inhibitor'])
+        create_report([
+            reporting.Title('Error while collecting system OS facts'),
+            reporting.Summary(str(e)),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Tags([reporting.Tags.OS_FACTS]),
```
Comment (shaded-enmity) [inline repos/system_upgrade/el7toel8/actors/ipuworkflowconfig/libraries/library.py]:
```suggestion
            reporting.Tags([reporting.Tags.OS_FACTS, reporting.Tags.SANITY]),
```

File: repos/system_upgrade/el7toel8/actors/migratentp/libraries/library.py [hunk 9]
```diff
@@ -80,12 +84,17 @@ def migrate_ntp(migrate_services, config_tgz64):
...
 
     if not ignored_lines:
-        reporting.report_generic(
-                title='{} configuration migrated to chrony'.format(' and '.join(migrate_configs)),
-                summary='ntp2chrony executed successfully',
-                severity='low')
+        create_report([
+            reporting.Title('{} configuration migrated to chrony'.format(' and '.join(migrate_configs))),
+            reporting.Summary('ntp2chrony executed successfully'),
+            reporting.Severity(reporting.Severity.LOW),
```
Comment (jmikovic) [inline repos/system_upgrade/el7toel8/actors/migratentp/libraries/library.py]:
```suggestion
            reporting.Severity(reporting.Severity.INFO),
```

File: repos/system_upgrade/el7toel8/actors/scheduleselinuxrelabeling/actor.py [hunk 10]
```diff
@@ -23,17 +26,21 @@ def process(self):
...
-                    report_generic(
-                        title='SElinux scheduled for relabelling',
-                        summary='/.autorelabel file touched on root in order to schedule SElinux relabelling.',
-                        severity='low',
-                    )
+                    create_report([
+                        reporting.Title('SElinux scheduled for relabelling'),
+                        reporting.Summary(
+                            '/.autorelabel file touched on root in order to schedule SElinux relabelling.'),
+                        reporting.Severity(reporting.Severity.LOW),
```
Comment (shaded-enmity) [inline repos/system_upgrade/el7toel8/actors/scheduleselinuxrelabeling/actor.py]:
```suggestion
                        reporting.Severity(reporting.Severity.INFO),
```

File: repos/system_upgrade/el7toel8/actors/scheduleselinuxrelabeling/actor.py [hunk 11]
```diff
@@ -23,17 +26,21 @@ def process(self):
...
-                        summary='./autorelabel file could not be created: {}.'.format(e),
-                        remediation='Please set autorelabelling manually after the upgrade.',
-                        severity='high'
-                    )
+                    create_report([
+                        reporting.Title('Could not schedule SElinux for relabelling'),
+                        reporting.Summary('./autorelabel file could not be created: {}.'.format(e)),
+                        reporting.Severity(reporting.Severity.HIGH),
+                        reporting.Tags(COMMON_REPORT_TAGS),
+                        reporting.Flags([reporting.Flags.FAILURE, reporting.Flags.INHIBITOR])
```
Comment (jmikovic) [inline repos/system_upgrade/el7toel8/actors/scheduleselinuxrelabeling/actor.py]: isn't it too late for inhibitor at this stage?

File: repos/system_upgrade/el7toel8/actors/scheduleselinuxrelabeling/actor.py [hunk 12]
```diff
@@ -23,17 +26,21 @@ def process(self):
...
+
                 except OSError as e:
                     # FIXME: add an "action required" flag later
-                    report_with_remediation(
-                        title='Could not schedule SElinux for relabelling',
-                        summary='./autorelabel file could not be created: {}.'.format(e),
-                        remediation='Please set autorelabelling manually after the upgrade.',
-                        severity='high'
-                    )
+                    create_report([
```
Comment (jmikovic) [inline repos/system_upgrade/el7toel8/actors/scheduleselinuxrelabeling/actor.py]: missing remediation

File: repos/system_upgrade/el7toel8/actors/setpermissiveselinux/actor.py [hunk 13]
```diff
@@ -25,10 +25,11 @@ def process(self):
...
-                        summary='{}'.format(err_msg),
-                        remediation='Please set SElinux into permissive mode manually.',
-                        severity='high',
-                    )
+                    create_report([
+                        reporting.Title('Could not set SElinux into permissive mode'),
+                        reporting.Summary('{}'.format(err_msg)),
+                        reporting.Severity(reporting.Severity.HIGH),
+                        reporting.Tags([reporting.Tags.SELINUX, reporting.Tags.SECURITY]),
+                        reporting.Flags([reporting.Flags.FAILURE, reporting.Flags.INHIBITOR])
```
Comment (jmikovic) [inline repos/system_upgrade/el7toel8/actors/setpermissiveselinux/actor.py]: again inhibitor in finalization phase

Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/setpermissiveselinux/actor.py]: yep, did not realized that back then. I think I made the same brain-fart even during the first reporting phase :)

File: repos/system_upgrade/el7toel8/actors/tcpwrapperscheck/actor.py [hunk 14]
```diff
@@ -54,13 +54,16 @@ def process(self):
...
-                              summary=('tcp_wrappers support has been removed in RHEL-8. '
-                                       'There is some configuration affecting installed packages (namely {}) '
-                                       'in /etc/hosts.deny or /etc/hosts.allow, which '
-                                       'is no longer going to be effective after update. '
-                                       'Please migrate it manually.'.format(', '.join(found_packages))),
-                              links=[{'title': 'Replacing TCP Wrappers in RHEL 8',
-                                      'href': 'https://access.redhat.com/solutions/3906701'}],
-                              severity='high',
-                              flags=['inhibitor'])
+            create_report([
```
Comment (jmikovic) [inline repos/system_upgrade/el7toel8/actors/tcpwrapperscheck/actor.py]: missing links

File: repos/system_upgrade/el7toel8/actors/scheduleselinuxrelabeling/actor.py (hunk 11 above)
Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/scheduleselinuxrelabeling/actor.py]: yep, did not realized that back then. I think I made the same brain-fart even during the first reporting phase :)

File: repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py (hunk 4 above)
Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/checkbtrfs/actor.py]: we add External link instead

File: repos/system_upgrade/el7toel8/actors/checksystemarch/actor.py (hunk 5 above)
Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/checksystemarch/actor.py]: Yep, makes sense. We should use this refactoring to improve/fix older reports...

File: repos/system_upgrade/el7toel8/actors/inhibitwhenluks/actor.py (hunk 7 above)
Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/inhibitwhenluks/actor.py]: fair enough

File: repos/system_upgrade/el7toel8/actors/authselectcheck/actor.py (hunk 2 above)
Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/authselectcheck/actor.py]: fixed

File: repos/system_upgrade/el7toel8/actors/tcpwrapperscheck/actor.py (hunk 14 above)
Comment (Rezney) [reply repos/system_upgrade/el7toel8/actors/tcpwrapperscheck/actor.py]: good catch

**Comment by Rezney** [comment]
@bocekm "verifycheckresults" is supposed to go:

https://github.com/oamg/leapp/pull/546

We decided to go this way just recently so I forgot to update this PR.

**Comment by bocekm** [comment]
Please create a Copr build so I can test it on Monday.

**Comment by Rezney** [comment]
All tests should be passing now...

**Comment by Rezney** [comment]
@bocekm fresh copr build created...

**Comment by jmikovic** [comment]
leapp-ci test

**Pull request merged**